compiler/rustc_ast_passes/src/ast_validation.rs

```diff
@@ -1277,8 +1289,16 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
         }
 
         let tilde_const_allowed =
-            matches!(fk.header(), Some(FnHeader { constness: ast::Const::Yes(_), .. }))
-                || matches!(fk.ctxt(), Some(FnCtxt::Assoc(_)));
+            if matches!(fk.header(), Some(FnHeader { constness: ast::Const::Yes(_), .. })) {
+                true
+            } else if let Some(FnCtxt::Assoc(ctxt)) = fk.ctxt() {
+                match ctxt {
+                    AssocCtxt::Trait => self.in_const_trait_defn,
+                    AssocCtxt::Impl => self.in_const_trait_impl,
+                }
+            } else {
+                false
+            };
```

Comment:
```suggestion
            } else {
                match fk.ctxt() {
                    Some(FnCtxt::Assoc(AssocCtxt::Trait)) => self.in_const_trait_defn,
                    Some(FnCtxt::Assoc(AssocCtxt::Impl)) => self.in_const_trait_impl,
                    _ => false,
                }
            };
```

Reply:
After combining `in_const_trait_defn` with `in_const_trait_impl`, I believe this can be further simplified/written more like the original code like (pushing new commit soon)